kep/48-node-resources-allocatable-scoring/README.md

```diff
@@ -0,0 +1,84 @@
+
+## Terms
+
+## Proposal
+
+In order to implement node resources allocatable scoring, we developed a score plugin. The plugin takes an arg ```mode```, which can have values ```Least``` or ```Most```. When ```Least``` is specified nodes with the least allocatable resources are scored higher. When ```Most``` is specified, nodes with the most allocatable resources are scored higher. ```Least``` and ```Most``` are totally opposite semantics, so it's only possible to have one enabled in a scheduler profile. The plugin takes another arg ```resources``` to specify the ```weight``` and ```name``` of each resource used in the scoring calculations.
+
+## Design Details
+We implement a score plugin based on the scheduler framework. Implementation follows a similar pattern to the existing ```NodeResourcesLeastAllocated``` score plugin. Each allocatable resource of a node is weighted and summed. After scores are calculated for all nodes, we normalize the scores. Here is a concrete example, comparing plugin ```NodeResourcesLeastAllocatable``` with ```NodeResourcesLeastAllocated```:
+
+Consider the following scenario:
+* 2 nodes with allocated/allocatable of 0/10, 0/200.
+* 3 pods are scheduled in the following order with the following requests: 5, 5, 100, 100
+
+With score plugin ```NodeResourcesLeastAllocatable```, all pods are scheduled.
+The final nodes have allocated/allocatable of 10/10, 200/200.
+
+With score plugin ```NodeResourceLeastAllocated```, one pod may be unscheduleable.
+The first pod may be placed on either node as both have equal ratio of allocated resources. Note that any scoring strategy that relies on ratio of allocated, will have this same issue. Even supposing the first pod landed on the smaller node, the second pod will be placed on the larger node as it has the least ratio of allocated resources. The final nodes have allocated/allocatable of 5/10, 105/200.
+
+Consider another scenario with churn:
+
+* 2 nodes with allocated/allocatable of 10/20 ( 1 pod: size 10), and 90/100 (2 pods: size 50 and 40)
+* 1 pod of size 10 comes, pod of size 40 is terminated, a new pod of size 50 is scheduled.
+
+With the ```NodeResourcesLeastAllocatable``` score plugin, all pods are scheduled. With other plugins, the size 10 pod may have been scheduled on the larger node and blocked the subsequent size 50 pod.
+
+### Extension points
+
+#### Score
+To score a node we take the sum of it's weighted allocatable resources. The resources and weights to use are user configurable via the plugin args. If the mode is ```least```, then we use the negation of the weighted sum. After all nodes are scored, the scores are normalized to a fixed range defined by the min and max node score in the framework (e.g. [0-100]).
```

Comment:
> If the mode is `least`

I suppose we're gonna use the uppercase `Least`, right?

Reply:
done